traversal/scanner/GraphIterator.java

```diff
@@ -301,200 +284,146 @@ private boolean verifyLoops() {
 
         private boolean isClosure(ProcedureEdge<?, ?> edge, Vertex<?, ?> fromVertex, Vertex<?, ?> toVertex) {
             if (edge.isRolePlayer()) {
-                Scopes.Scoped scoped = scopes.getOrInitialise(edge.asRolePlayer().scope());
-                return edge.asRolePlayer().isClosure(graphMgr, fromVertex, toVertex, params, scoped);
+                Scope scope = scopes.get(edge.asRolePlayer().scope());
+                return edge.asRolePlayer().isClosure(graphMgr, fromVertex, toVertex, params, scope);
             } else {
                 return edge.isClosure(graphMgr, fromVertex, toVertex, params);
             }
         }
 
-        private void addEdgeInput(ProcedureEdge<?, ?> edge, Vertex<?, ?> fromVertex) {
-            assert !edgeInputs.containsKey(edge);
-            if (iterator != null && iterator.hasNext()) {
-                edgeInputIntersectionCache.put(new HashMap<>(edgeInputs), iterator.peek());
-                iterator = intersect(branch(fromVertex, edge), iterator);
-            }
-            edgeInputs.put(edge, fromVertex);
-            vertex = null;
-        }
-
-        private void removeEdgeInput(ProcedureEdge<?, ?> edge) {
-            edgeInputIntersectionCache.remove(edgeInputs);
-            edgeInputs.remove(edge);
-            reset();
-        }
```

Comment:
we no longer have to "activate" and "deactivate" different input edges to use when traverse this vertex - by the time we arrive at traversing a vertex, all predecessors/dependees must have been explored & valid! This is much simpler to understand